sky/sdk/lib/widgets/widget.dart

```diff
@@ -44,10 +44,86 @@ class ObjectKey extends Key {
   int get hashCode => identityHashCode(value);
 }
 
+typedef void GlobalKeyRemovalListener(GlobalKey key);
+
 abstract class GlobalKey extends Key {
   GlobalKey.constructor() : super.constructor(); // so that subclasses can call us, since the Key() factory constructor shadows the implicit constructor
   factory GlobalKey({ String label }) => new LabeledGlobalKey(label);
   factory GlobalKey.fromObjectIdentity(Object value) => new GlobalObjectKey(value);
+
+  static final Map<GlobalKey, Widget> _registry = new Map<GlobalKey, Widget>();
+  static final Map<GlobalKey, int> _debugDuplicates = new Map<GlobalKey, int>();
+  static final Map<GlobalKey, Set<GlobalKeyRemovalListener>> _removalListeners = new Map<GlobalKey, Set<GlobalKeyRemovalListener>>();
+  static final Set<GlobalKey> _removedKeys = new Set<GlobalKey>();
+
+  void _register(Widget widget) {
+    assert(() {
+      if (_registry.containsKey(this)) {
+        int oldCount = _debugDuplicates.putIfAbsent(this, () => 1);
+        assert(oldCount >= 1);
+        _debugDuplicates[this] = oldCount + 1;
+      }
+      return true;
+    });
+    _registry[this] = widget;
+  }
+
+  void _unregister(Widget widget) {
+    assert(() {
+      if (_registry.containsKey(this) && _debugDuplicates.containsKey(this)) {
+        int oldCount = _debugDuplicates[this];
+        assert(oldCount >= 2);
+        if (oldCount == 2) {
+          _debugDuplicates.remove(this);
+        } else {
+          _debugDuplicates[this] = oldCount - 1;
+        }
+      }
+      return true;
+    });
+    if (_registry[this] == widget) {
+      _registry.remove(this);
+      _removedKeys.add(this);
+    }
+  }
+
+  static bool _notifyingListeners = false;
+
+  static void registerRemovalListener(GlobalKey key, GlobalKeyRemovalListener listener) {
+    assert(!_notifyingListeners);
+    assert(key != null);
+    if (!_removalListeners.containsKey(key))
+      _removalListeners[key] = new Set<GlobalKeyRemovalListener>();
+    bool added = _removalListeners[key].add(listener);
+    assert(added);
+  }
+
+  static void unregisterRemovalListener(GlobalKey key, GlobalKeyRemovalListener listener) {
+    assert(!_notifyingListeners);
+    assert(key != null);
+    assert(_removalListeners.containsKey(key));
+    bool removed = _removalListeners[key].remove(listener);
+    if (_removalListeners[key].isEmpty)
+      _removalListeners.remove(key);
+    assert(removed);
+  }
+
+  static void _notifyListeners() {
+    assert(!_inRenderDirtyComponents);
+    assert(!Widget._notifyingMountStatus);
+    assert(_debugDuplicates.isEmpty);
+    _notifyingListeners = true;
+    for (GlobalKey key in _removedKeys) {
+      if (!_registry.containsKey(key) && _removalListeners.containsKey(key)) {
+        for (GlobalKeyRemovalListener listener in _removalListeners[key])
+          listener(key);
+        _removalListeners.remove(key);
+      }
+    }
+    _removedKeys.clear();
+    _notifyingListeners = false;
```

Comment:
try { ... } finally { ... } in case listener throws, we don't want to screw up our state machine.